OWML.ModHelper.Input/ModInputHandler.cs

```diff
@@ -0,0 +1,338 @@
+            UpdateCombo();
+            return _lastPressed == combination;
+        }
+
+        public bool IsNewlyPressedExact(IModInputCombination combination, bool keep = false)
+        {
+            UpdateCombo();
+            return _lastPressed == combination && combination.IsFirst(keep);
+        }
+
+        public bool WasTappedExact(IModInputCombination combination, bool keep = false)
+        {
+            UpdateCombo();
+            return combination != _lastPressed
+                && (combination.PressDuration < TapDuration)
+                && combination.IsFirst(keep);
+        }
+
+        public bool WasNewlyReleasedExact(IModInputCombination combination, bool keep = false)
+        {
+            UpdateCombo();
+            return _lastPressed != combination && combination.IsFirst(keep);
+        }
+
+        private bool IsPressedSingle(IModInputCombination combination)
+        {
+            if (Time.realtimeSinceStartup - _lastSingleUpdate > Time.unscaledDeltaTime / 2)
+            {
+                _lastSingleUpdate = Time.realtimeSinceStartup;
+                var toRemove = new List<IModInputCombination>();
+                foreach (var combo in _singlePressed)
+                {
+                    if (!IsPressed(combo))
+                    {
+                        ((ModInputCombination)combo).SetPressed(false);
+                        toRemove.Add(combo);
+                    }
+                }
+                foreach (var combo in toRemove)
+                {
+                    _singlePressed.Remove(combo);
+                }
+            }
+            foreach (var key in ((ModInputCombination)combination).GetSingles())
+            {
+                if (UnityEngine.Input.GetKey(key) && !ShouldIgnore(key))
+                {
+                    _singlePressed.Add(combination);
+                    ((ModInputCombination)combination).SetPressed();
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public bool IsPressed(IModInputCombination combination)
+        {
+            return IsPressedSingle(combination) || IsPressedExact(combination);
+        }
+
+        public bool IsNewlyPressed(IModInputCombination combination, bool keep = false)
+        {
+            return IsPressed(combination) && combination.IsFirst(keep);
+        }
+
+        public bool WasTapped(IModInputCombination combination, bool keep = false)
+        {
+            return (!(IsPressedExact(combination) || IsPressedSingle(combination)))
+                && (combination.PressDuration < TapDuration) 
+                && combination.IsFirst(keep);
+        }
+
+        public bool WasNewlyReleased(IModInputCombination combination, bool keep = false)
+        {
+            return (!(IsPressedExact(combination) || IsPressedSingle(combination)))
+                && combination.IsFirst(keep);
+        }
+
+        private KeyCode StringToKeyCode(string key)
```

Comment:
This method is a bit complicated to read. I think it would be nice if you extract the contents into smaller methods with intuitive names.

Reply:
separated part that handles keyboard keys